### PR TITLE
[NPROD-80] API export update

### PR DESCRIPTION
**Pull request by F3kilo**
This PR improves integration of the IS20 API to other canisters with the following features:
- Optional export of the mint, burn and transfer methods. Export of these methods can be disabled with corresponding features.
- The `inspect_message` call separated into the inspection logic and the exported API. The API is exported only from the `is20-token-canister` package. The `is20-token` package contains inspection logic that can be used in the `inspect_message` calls of other canister, that provides IS20 API.

### CODE REVIEW

**Comment by blutooth** [comment]
Are these  sorts of attributes #[cfg_attr(not(feature = "no_mint_burn_api"), necessary? I don't think so because use of traits allows you to override the default implementation with empty methods rather than not importing with a macro?

**Comment by F3kilo** [comment]
> Are these sorts of attributes #[cfg_attr(not(feature = "no_mint_burn_api"), necessary? I don't think so because use of traits allows you to override the default implementation with empty methods rather than not importing with a macro?

In the Neuro we need these methods for internal use (probably overloaded version), but we don't want to provide public access to them. So @Maximkaaa suggests this approach to disable export.

**Comment by awkure** [comment]
I don't think we should accumulate features as well. To prevent calling canister methods there're three ways 
1) Set up `PreUpdate` call in your local canister to filter out necessary calls
2) Overwrite default trait implementation of an endpoint with an empty body 
3) Add a macro attribute to the `canister-sdk` macro api something like `#[do_not_export]`, implement the feature with empty body as in `2` and probably filter them out when merging idls. I'm not sure if the method with empty body will ever be exported in first place.

Regarding the `inspect_message` I think It'll be better if we do not export it as an endpoint (because semantically it's not, but something that runs *before* the call), but if you want to inherit this validation logic in your local canister, you should call it as a function in `inspect_message` body of your canister.

**Comment by Maximkaaa** [comment]
@blutooth The thing is that only pure methods of the canister traits can be overloaded. This is the limitation of the approach we take to export these methods. E.g. if the method is exported by the trait, it's exported as declared in the trait. So if we want to overload it somewhere else, we must not export in in the trait, but rather provide the default implementation in the trait and then export it in the canister that implements the trait. See, for example, how the `upgrade` method in the factories works. The `FactoryCanister` trait does not export this method. Instead, it provides the implementation, and then the factory structs declare the API exports, in which they call this method (it must be overloaded because of the logic to get canister wasm).

So, if we don't want to use rust features to opt-out the exports, we would have to not export these methods (the methods that we want to overload sometimes) and make the canister struct declare the exports every time. With features we can reduce the boilerplate code amount.

**Comment by blutooth** [comment]
I think the trait + struct approach would be preferable, as it provides a template for doing this. With macros we have to write custom stuff which is not desirable. I wuld prefer the boilerplate in this case.

**Review by blutooth — APPROVED** [review]

File: src/token/api/Cargo.toml [hunk 1]
```diff
@@ -4,8 +4,10 @@ name = "is20-token"
 version = "0.1.7"
 
 [features]
-default = []
+default = ["mint_burn", "transfer"]
 no_api = []
+mint_burn = []
+transfer = []
```
Comment (Maximkaaa) [inline src/token/api/Cargo.toml:10]: Let's add comments about these features

File: src/token/api/src/canister/inspect.rs [hunk 2]
```diff
@@ -49,53 +48,55 @@ static TRANSACTION_METHODS: &[&str] = &[
...
+    state: &CanisterState,
+    method: &str,
+    caller: Principal,
+) -> Result<AcceptReason, &'static str> {
+    match method {
         // These are query methods, so no checks are needed.
-        "mint" if state.stats.is_test_token => ic_cdk::api::call::accept_message(),
-        m if PUBLIC_METHODS.contains(&m) => ic_cdk::api::call::accept_message(),
+        "mint" if state.stats.is_test_token => Ok(AcceptReason::Valid),
+        m if PUBLIC_METHODS.contains(&m) => Ok(AcceptReason::Valid),
```
Comment (Maximkaaa) [inline src/token/api/src/canister/inspect.rs:72]: If we use exclude these methods from the API by using the features, we probably don't wont the logic for their inspection to be here also.

**Review by Maximkaaa — CHANGES_REQUESTED** [review]

File: src/token/api/src/canister/inspect.rs (hunk 2 above)
Comment (blutooth) [reply src/token/api/src/canister/inspect.rs:73]: would be nice in that case if the inspect message methods could be imported separately and compiled as one.

**Review by blutooth — APPROVED** [review]

File: src/token/api/src/canister/inspect.rs (hunk 2 above)
Comment (F3kilo) [reply src/token/api/src/canister/inspect.rs:73]: Inspections are disabled for the excluded methods now.

**Pull request merged**